File: 2019/day05/cpp/main.cpp
#include <cstdlib>

#include <algorithm>
#include <iostream>
#include <fstream>
#include <vector>
#include <array>
#include <tuple>
#include <stdexcept>

enum class ParameterMode {
  UNKNOWN,
  POSITION,
  IMMEDIATE,
};

std::tuple<std::array<ParameterMode, 3>, int> parse_opcode(int instruction) {
  const auto opcode = instruction % 100;
  std::array<ParameterMode, 3> modes{};

  instruction /= 100;
  modes[0] = instruction % 10 == 0 ? ParameterMode::POSITION : ParameterMode::IMMEDIATE;
  instruction /= 10;
  modes[1] = instruction % 10 == 0 ? ParameterMode::POSITION : ParameterMode::IMMEDIATE;
  instruction /= 10;
  modes[2] = instruction % 10 == 0 ? ParameterMode::POSITION : ParameterMode::IMMEDIATE;

  if (std::any_of(modes.begin(), modes.end(),
    [](auto mode){ return mode == ParameterMode::UNKNOWN; })) {
    throw std::runtime_error{"Unknown parameter mode."};
  }

  return std::make_tuple(modes, opcode);
}

void run_program(std::vector<int> program) {
  std::size_t rip{};
  for (;;) {
    const auto [modes, opcode] = parse_opcode(program[rip]);
    switch (opcode) {
      case 1: {
        const auto value_1 = modes[0] == ParameterMode::POSITION ?
          program[program[rip + 1]] : program[rip + 1];
        const auto value_2 = modes[1] == ParameterMode::POSITION ?
            program[program[rip + 2]] : program[rip + 2];
        // ignore parameters mode for the output
        program[program[rip + 3]] = value_1 + value_2;
        rip += 4;
        break;
      }

      case 2: {
        const auto value_1 = modes[0] == ParameterMode::POSITION ?
          program[program[rip + 1]] : program[rip + 1];
        const auto value_2 = modes[1] == ParameterMode::POSITION ?
            program[program[rip + 2]] : program[rip + 2];
        // ignore parameters mode for the output
        program[program[rip + 3]] = value_1 * value_2;
        rip += 4;
        break;
      }


      case 3: {
        int value;
        std::cin >> value;
        // ignore the parameters mode for output address
        program[program[rip + 1]] = value;
        rip += 2;
        break;
      }

      case 4:
        if (modes[0] == ParameterMode::POSITION) {
          std::cout << program[program[rip + 1]] << std::endl;
        }
        if (modes[0] == ParameterMode::IMMEDIATE) {
          std::cout << program[rip + 1] << std::endl;
        }
        rip += 2;
        break;

      case 99:
        return;

      default:
        throw std::runtime_error{"Read unknown opcode."};
    }
  }
}

int main() {
  std::ifstream input_file{"./input.txt"};
  if (not input_file.good()) {
    std::cerr << "Could not open the input file.\n";
    return 1;
  }

  std::vector<int> program;
  int value;
  while(input_file >> value) {
    program.push_back(value);
    if (input_file.peek() == ',')
      input_file.get();
  }

  run_program(program);

  return 0;
}

File: 2019/day05/python/main.py
#!/usr/bin/python3

with open('input.txt', 'r') as f:
  input = f.read()

program = [int(ins) for ins in input.split(',')]

POSITION = 0
IMMEDIATE = 1

def decode_opcode(ins):
  modes = []
  opcode = ins % 100
  ins = ins // 100
  modes.append(ins % 10)
  ins = ins // 10
  modes.append(ins % 10)
  ins = ins // 10
  modes.append(ins % 10)

  return [modes, opcode]


def run_program(program, value):
  rip = 0
  while True:
    modes, opcode = decode_opcode(program[rip])
    if opcode == 1:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      value_2 = program[program[rip + 2]] if modes[1] == POSITION else program[rip + 2]
      # ignore parameter mode for the output
      program[program[rip + 3]] = value_1 + value_2
      rip += 4
    elif opcode == 2:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      value_2 = program[program[rip + 2]] if modes[1] == POSITION else program[rip + 2]
      # ignore parameter mode for the output
      program[program[rip + 3]] = value_1 * value_2
      rip += 4
    elif opcode == 3:
      # ignore parameter output for the output
      program[program[rip + 1]] = value
      rip += 2
    elif opcode == 4:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      print(value_1)
      rip += 2
    elif opcode == 5:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      value_2 = program[program[rip + 2]] if modes[1] == POSITION else program[rip + 2]
      if value_1 != 0:
        rip = value_2
      else:
        rip += 3
    elif opcode == 6:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      value_2 = program[program[rip + 2]] if modes[1] == POSITION else program[rip + 2]
      if value_1 == 0:
        rip = value_2
      else:
        rip += 3
    elif opcode == 7:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      value_2 = program[program[rip + 2]] if modes[1] == POSITION else program[rip + 2]
      # ignore parameter output for the output
      program[program[rip + 3]] = 1 if value_1 < value_2 else 0
      rip += 4
    elif opcode == 8:
      value_1 = program[program[rip + 1]] if modes[0] == POSITION else program[rip + 1]
      value_2 = program[program[rip + 2]] if modes[1] == POSITION else program[rip + 2]
      # ignore parameter output for the output
      program[program[rip + 3]] = 1 if value_1 == value_2 else 0
      rip += 4
    elif opcode == 99:
      return
    else:
      raise Exception("Unknown opcode %d." % opcode)


print("=============[Stage1]==============: \n\n")
run_program(program.copy(), 1)
print("\n=============[Stage2]==============: \n\n")
run_program(program.copy(), 5)